packages/controllers/src/permissions/PermissionController.ts

```diff
@@ -0,0 +1,1922 @@
+    };
+
+    const permissionsRequest: PermissionsRequest = {
+      metadata,
+      permissions: requestedPermissions,
+    };
+
+    const { permissions: approvedPermissions, ...requestData } =
+      await this.requestUserApproval(permissionsRequest);
+    if (Object.keys(approvedPermissions).length === 0) {
+      // The request should already have been rejected if this is the case.
+      throw internalError(
+        `Approved permissions request for origin "${origin}" contains no permissions.`,
+        { requestedPermissions },
+      );
+    }
+
+    return [
+      this.grantPermissions({
+        subject: { origin },
+        approvedPermissions,
+        preserveExistingPermissions,
+        requestData,
+      }),
+      metadata,
+    ];
+  }
+
+  /**
+   * Adds a request to the {@link ApprovalController} using the
+   * {@link AddApprovalRequest} action.
+   *
+   * @param permissionsRequest - The permissions request object.
+   * @returns The approved permissions request object.
+   */
+  private async requestUserApproval(permissionsRequest: PermissionsRequest) {
+    const { origin, id } = permissionsRequest.metadata;
+    return (await this.messagingSystem.call(
+      'ApprovalController:addRequest',
+      {
+        id,
+        origin,
+        requestData: permissionsRequest,
+        type: MethodNames.requestPermissions,
+      },
+      true,
+    )) as PermissionsRequest;
+  }
+
+  /**
+   * Validates requested permissions. Throws if validation fails.
+   *
+   * All properties of {@link RequestedPermissions} values are ignored during
+   * request processing in this controller except `caveats`, which is validated
+   * by {@link PermissionController.computeCaveats}, caveat validators, and
+   * permission validators.
+   *
+   * This method still validates the `parentCapability` field, since a mismatch
+   * between a {@link RequestedPermissions} key and `parentCapability` makes it
+   * impossible to determine which permission the consumer intended to request.
+   * At the time of writing, our types also do not prevent this kind of
+   * mismatch.
+   *
+   * @param origin - The origin of the grantee subject.
+   * @param requestedPermissions - The requested permissions.
+   */
+  private validateRequestedPermissions(
+    origin: OriginString,
+    requestedPermissions: RequestedPermissions,
+  ): void {
+    if (
+      !requestedPermissions ||
+      typeof requestedPermissions !== 'object' ||
+      Array.isArray(requestedPermissions) ||
+      Object.keys(requestedPermissions).length === 0
+    ) {
+      throw invalidParams({ data: { origin, requestedPermissions } });
+    }
+
+    const perms: RequestedPermissions = requestedPermissions;
+
+    for (const methodName of Object.keys(perms)) {
+      const target = (
+        perms[methodName] as unknown as ExtractPermission<
+          ControllerPermissionSpecification,
+          ControllerCaveatSpecification
+        >
+      ).parentCapability;
```

Comment:
This cast appears to be unnecessary
```suggestion
      const target = perms[methodName].parentCapability;
```